src/rust/catmem/mod.rs

```diff
@@ -110,318 +101,199 @@ impl CatmemLibOS {
-                    })
-                } else {
-                    let qtable_ptr: Rc<RefCell<IoQueueTable<CatmemQueue>>> = self.qtable.clone();
-                    Box::pin(async move {
-                        // Wait for pop to complete.
-                        let result: Result<(DemiBuffer, bool), Fail> = pop_coroutine(ring, size, yielder).await;
-                        // Process the result.
-                        match result {
-                            Ok((buf, eof)) => {
-                                if eof {
-                                    let mut qtable_: RefMut<IoQueueTable<CatmemQueue>> = qtable_ptr.borrow_mut();
-                                    let queue: &mut CatmemQueue = match qtable_.get_mut(&qd) {
-                                        Some(queue) => queue,
-                                        None => {
-                                            let cause: String = format!("invalid queue descriptor (qd={:?})", qd);
-                                            error!("pop(): {}", cause);
-                                            return (qd, OperationResult::Failed(Fail::new(libc::EBADF, &cause)));
-                                        },
-                                    };
-                                    let pipe: &mut Pipe = queue.get_mut_pipe();
-                                    pipe.set_eof();
-                                }
-                                (qd, OperationResult::Pop(buf))
-                            },
-                            Err(e) => (qd, OperationResult::Failed(e)),
-                        }
-                    })
-                };
-
-                let task_id: String = format!("Catmem::pop for qd={:?}", qd);
-                let task: OperationTask = OperationTask::new(task_id, coroutine);
-                let handle: TaskHandle = match self.scheduler.insert(task) {
-                    Some(handle) => handle,
+        let queue: CatmemQueue = self.get_queue(qd)?;
+        let yielder: Yielder = Yielder::new();
+        let yielder_handle: YielderHandle = yielder.get_handle();
+        let qtable_ptr: Rc<RefCell<IoQueueTable<CatmemQueue>>> = self.qtable.clone();
+        let coroutine: Pin<Box<Operation>> = Box::pin(async move {
+            // Wait for pop to complete.
+            let (buf, eof) = match queue.pop(size, yielder).await {
+                Ok((buf, eof)) => (buf, eof),
+                Err(e) => return (qd, OperationResult::Failed(e)),
+            };
+
+            let mut qtable_: RefMut<IoQueueTable<CatmemQueue>> = qtable_ptr.borrow_mut();
+
+            if eof {
+                match qtable_.get_mut(&qd) {
+                    Some(queue) => {
+                        // return (
+                        //     qd,
+                        //     OperationResult::Failed(Fail::new(libc::ECONNRESET, "this queue was closed")),
+                        // );
```

Comment:
Should this be commented out?

Reply:
Ah, yes it should. It does not comply with our current API spec, which is to return 0 bytes on the first pop with EOF set and then ECONNRESET on subsequent ones. If I uncomment that it will break the tests.

Reply:
Fixed.